pulpcore/cli/common/generic.py

```diff
@@ -328,59 +328,52 @@ def _version_callback(
-# to build "on top of" it like I'm doing now w/ json_callback
 def load_file_or_string_callback(
-    ctx: click.Context, param: click.Parameter, value: Optional[str]
+    handler: Callable[[click.Context, click.Parameter, str], Any] = lambda c, p, x: x
 ) -> Any:
-    """Load string from input or from file if string starts with @."""
-    the_content: str
+    @wraps(handler)
+    def _load_file_or_string_callback(
+        ctx: click.Context, param: click.Parameter, value: Optional[str]
+    ) -> Any:
+        """Load the string from input, or from file if the value starts with @."""
+        if not value:
+            return value
 
-    # pass None and "" verbatim
-    if not value:
-        return value
+        if value.startswith("@"):
+            the_file = value[1:]
+            try:
+                with click.open_file(the_file, "r") as fp:
+                    the_content = fp.read()
+            except OSError:
+                raise click.ClickException(
+                    _("Failed to load content from {file}").format(file=the_file)
+                )
+        else:
+            the_content = value
 
-    if value.startswith("@"):
-        the_file = value[1:]
-        try:
-            with click.open_file(the_file, "r") as fp:
-                the_content = fp.read()
-        except OSError:
-            raise click.ClickException(
-                _("Failed to load content from {file}").format(file=the_file)
-            )
-    else:
-        the_content = value
+        return handler(ctx, param, the_content)
 
-    return the_content
+    return _load_file_or_string_callback
 
 
-def load_json_callback(ctx: click.Context, param: click.Parameter, value: Optional[str]) -> Any:
-    """Load JSON from input string or from file if string starts with @."""
+@load_file_or_string_callback
+def string_callback(ctx: click.Context, param: click.Parameter, value: str) -> str:
+    return value
 
-    # None or empty-str are legal - shortcircuit here
-    if not value:
-        return value
 
-    # Now try to evaluate legal JSON
-    json_object: Any
-    json_string: str = load_file_or_string_callback(ctx, param, value)
+@load_file_or_string_callback
+def json_callback(ctx: click.Context, param: click.Parameter, value: str) -> Any:
     try:
-        json_object = json.loads(json_string)
+        json_object = json.loads(value)
     except json.decoder.JSONDecodeError:
         raise click.ClickException(_("Failed to decode JSON"))
     else:
         return json_object
 
 
-def labels_callback(
-    ctx: click.Context, param: click.Parameter, value: Optional[str]
-) -> Optional[Dict[str, str]]:
-    # None is legal - shortcircuit here
-    if value is None:
-        return value
-
-    value = load_json_callback(ctx, param, value)
+@load_file_or_string_callback
+def labels_callback(ctx: click.Context, param: click.Parameter, value: str) -> Any:
+    value = json_callback(ctx, param, value)
```

Comment:
Maybe `json_callback` should follow the chaining pattern. Also this version of `json_callback` is already decorated. So we are attemting to lookup a file twice here.

Reply:
I think, i'd be ok to use the load_json_callback here. We never want a version of the label callback that cannot read from file, right?

Reply:
Makes perfect sense. Thanks!